cpp/src/arrow/compute/kernels/scalar_arithmetic.cc

```diff
@@ -451,8 +655,49 @@ struct ArithmeticFunction : ScalarFunction {
     if (auto kernel = DispatchExactImpl(this, *values)) return kernel;
     return arrow::compute::detail::NoMatchingKernel(this, *values);
   }
+
+  Status CheckDecimals(std::vector<ValueDescr>* values) const {
+    bool has_decimal = false;
+    for (const auto& value : *values) {
+      if (is_decimal(value.type->id())) {
+        has_decimal = true;
+        break;
+      }
+    }
+    if (!has_decimal) return Status::OK();
+
+    if (values->size() == 2) {
+      return CastBinaryDecimalArgs(name(), values);
+    }
+    return Status::OK();
+  }
+};
+
+// resolve decimal operation output type
+struct BinaryDecimalOutputResolver {
+  const std::string func_name;
+
+  explicit BinaryDecimalOutputResolver(std::string func_name)
+      : func_name(std::move(func_name)) {}
+
+  Result<ValueDescr> operator()(KernelContext*, const std::vector<ValueDescr>& args) {
+    ARROW_ASSIGN_OR_RAISE(auto type, ResolveBinaryDecimalOutput(func_name, args));
+    return ValueDescr(std::move(type), GetBroadcastShape(args));
+  }
 };
 
+template <typename Op>
+void AddDecimalBinaryKernels(const std::string& name,
+                             std::shared_ptr<ArithmeticFunction>* func) {
+  auto out_type = OutputType(BinaryDecimalOutputResolver(name));
```

Comment:
I think it'd be more straightforward to dispatch on name here:
```suggestion
  OutputType out_type;
  const std::string op = func_name.substr(0, func_name.find("_"));
  if (op == "add" || op == "subtract") {
    out_type = OutputType(ResolveDecimalAdditionOrSubtractionOutput);
  } else if (op == "multiply") {
    out_type = OutputType(ResolveDecimalMultiplicationOutput);
  } else if (op == "divide") {
    out_type = OutputType(ResolveDecimalDivisionOutput);
  } else {
    DCHECK(false);
  }
```

Reply:
Done